src/utils/generator.js

```diff
@@ -1253,3 +1325,108 @@ export class PreludeGenerator {
     return ref;
   }
 }
+
+type TemporalBuildNodeEntryOptimizationStatus = "NO_OPTIMIZATION" | "POSSIBLE_OPTIMIZATION";
+
+// This function attempts to optimize Object.assign calls, by merging mulitple
+// calls into one another where possible. For example:
+//
+// var a = Object.assign({}, someAbstact);
+// var b = Object.assign({}, a);
+//
+// Becomes:
+// var b = Object.assign({}, someAbstract, a);
+//
+export function attemptToMergeEquivalentObjectAssigns(
+  realm: Realm,
+  callbacks: VisitEntryCallbacks,
+  temporalBuildNodeEntry: TemporalBuildNodeEntry
+): TemporalBuildNodeEntryOptimizationStatus | TemporalObjectAssignEntry {
+  let args = temporalBuildNodeEntry.args;
+  // If we are Object.assigning 3 or more args
+  if (args.length < 3) {
+    return "NO_OPTIMIZATION";
+  }
+  let objectAssingFunc = args[0];
+  let to = args[1];
+  // Then scan through the args after the "to" of this Object.assign, to see if any
+  // other sources are the "to" of a previous Object.assign call
+  loopThroughArgs: for (let i = 2; i < args.length; i++) {
+    let possibleOtherObjectAssignTo = args[i];
+    // Ensure that the "to" value can be omitted
+    if (!callbacks.canOmit(possibleOtherObjectAssignTo)) {
+      continue;
+    }
+    // Check if the "to" was definitely an Object.assign, it should
+    // be a snapshot AbstractObjectValue
+    if (possibleOtherObjectAssignTo instanceof AbstractObjectValue) {
+      let otherTemporalBuildNodeEntry = realm.getTemporalBuildNodeEntryFromDerivedValue(possibleOtherObjectAssignTo);
+      if (otherTemporalBuildNodeEntry === undefined) {
+        continue;
+      }
+      let otherArgs = otherTemporalBuildNodeEntry.args;
+      // Object.assign has at least 2 args and the first is that of Object.assign native function
+      if (otherArgs.length < 3 || otherArgs[0] !== objectAssingFunc) {
+        continue;
+      }
+      let otherArgsToUse = [];
+      for (let x = 2; x < otherArgs.length; x++) {
+        let arg = otherArgs[x];
```

Comment:
If the `arg` is an `ObjectValue`, you must also check if it `mightBeHavocedObject`, and then you cannot optimize when something that may be an abstract function call happened in between `startIndex` and `endIndex`. And this needs a test case. @sebmarkbage knows what this is all about, and @sb98052 is on a quest to refactor all of that, including proper accounting for aliasing.

Reply:
Didn't run it, but something like this:

```
(function () {
  function f(o1, o2, g, h) {
    g(o1, o2);
    let p = Object.assign({}, o1);
    h(o2); // can mutate o1 !
    let q = Object.assign({}, p);
    return q;
  }
  __optimize(f);
  inspect = function() {
    return f({}, {}, 
      function(o1, o2) {
        o2.o1 = o1;
      },
      function(o2) {
        o2.o1.x = 42;
      }
    ).x;
  }
})();
```

Reply:
So I added logic to do what you said above, but the issue is that `o1` can never be marked as havoc as it's never an `AbstractObjectValue` or `ObjectValue`. It sounds like we're doing something here that we don't currently support in Prepack, so I'm not sure how to model it. Regardless, I'll add that logic you mentioned.

Reply:
This test works with your change, and works with my previous concerns:

```js
function f(o1, o2, g, h) {
  let a = Object.assign({}, o1);
  g(a, o2);
  let p = Object.assign({}, a);
  h(o2); // can mutate a !
  let q = Object.assign({}, p);
  return q;
}
```

Reply:
@NTillmann Your last test passes with the changes in this PR. :)